stan/math/opencl/copy.hpp

```diff
@@ -108,14 +124,19 @@ inline std::vector<double> packed_copy(const matrix_cl& src) {
   if (dst.size() == 0) {
     return dst;
   }
-  cl::CommandQueue queue = opencl_context.queue();
   try {
+    cl::CommandQueue queue = opencl_context.queue();
     matrix_cl packed(packed_size, 1);
     stan::math::opencl_kernels::pack(cl::NDRange(src.rows(), src.rows()),
-                                     packed.buffer(), src.buffer(), src.rows(),
-                                     src.rows(), triangular_view);
-    queue.enqueueReadBuffer(packed.buffer(), CL_TRUE, 0,
-                            sizeof(double) * packed_size, dst.data());
+                                     packed, src, src.rows(), src.rows(),
+                                     triangular_view);
+    auto mat_events = vec_concat(packed.read_events(), src.write_events());
+    cl::Event copy_event;
+    queue.enqueueReadBuffer(packed.buffer(), CL_FALSE, 0,
+                            sizeof(double) * packed_size, dst.data(),
+                            &mat_events, &copy_event);
+    copy_event.wait();
+    src.clear_write_events();
```

Comment:
Here I'd actually prefer removing the `copy_event.wait()` and not clearing any events so that the async keeps going. I think we should only clear things out when its `CPU -> OpenCL` or `OpenCL -> CPU`

Reply:
Yes since packed is waiting on `src` read events and the `enqueueReadBuffer` waits on the write events then that makes sense